chameleon/views.py

```diff
@@ -1,17 +1,106 @@
 from django.conf import settings
 from django.core.urlresolvers import reverse
 from django.contrib.auth.decorators import login_required
-from django.shortcuts import render
+from django.shortcuts import render, redirect
 from user_news.models import Outage
 from djangoRT import rtUtil
-
+from urlparse import urlparse
 import logging
+import chameleon.os_login as login
+from tas import auth as tas_auth
 from pytas.models import Project
 from webinar_registration.models import Webinar
 from django.utils import timezone
+from django.core.exceptions import SuspiciousOperation
+import sys
+from django.contrib.auth.forms import AuthenticationForm
+from django.conf import settings
+from django.http import HttpResponseRedirect
 
 logger = logging.getLogger(__name__)
 
+@login_required
+def horizon_sso_login(request):
+    next = ''
+    ## first we get the url params host, webroot, and next
+    horizon_webroot = request.GET.get('webroot') if request.GET.get('webroot') else ''
+    next = request.GET.get('next') if request.GET.get('next') else ''
+    host = request.GET.get('host') if request.GET.get('host') else ''
+
+    ## now we verify the host is valid, if not valid, log the error, send to home page...
+    valid_callback_hosts = getattr(settings, 'SSO_CALLBACK_VALID_HOSTS', [])
+    if not host or not host in valid_callback_hosts:
+        logger.error('Invalid or missing host in callback by user ' \
+            + request.user.username + ', callback host was: ' + host)
+        return HttpResponseRedirect('/')
+
+    ## get the Keystone token, and add to context 
+    ## if none available, prompt user for username/password to get one
+    context = {}
+    try:
+        context['sso_token'] = request.session['unscoped_token'].get('auth_token')
+    except:
+        return manual_ks_login(request)
```

Comment:
Kind of not ideal that we have to maintain 2 separate login pages now, but I get why... the login form supports other methods of login, like GENI. Actually... I wonder how being logged in via GENI or something would affect this flow. (Do people login with GENI in practice?)

Reply:
Agree, this won't work for login, it doesn't create a django session and to get here we actually need to be logged in. The form allows users who either logged in with Geni or who's account was in a transient state at original login to retrieve a Keystone token. Both are edge cases, but since the portal has long sessions and users may not log out and back in after creating their portal account, this gives them a way to either get a token or get an informational message about why they can't log in to Horizon.